components/msg/background_hang_monitor.rs

```diff
@@ -2,50 +2,244 @@
+unsafe fn suspend_thread(thread_id: MonitoredThreadId, success: &mut bool) {
+    match libc::pthread_kill(thread_id, libc::SIGPROF) {
+        0 => {},
+        _ => {
+            *success = false;
+        },
+    };
+}
+
+#[cfg(target_os = "macos")]
+#[allow(unsafe_code)]
+unsafe fn get_registers(thread_id: MonitoredThreadId, regs: &mut Registers, success: &mut bool) {
+    let state = mach::structs::x86_thread_state64_t::new();
+    let state_count = mach::structs::x86_thread_state64_t::count();
+    let kret = mach::thread_act::thread_get_state(thread_id,
+                                                  mach::thread_status::x86_THREAD_STATE64,
+                                                  mem::transmute(&state),
+                                                  mem::transmute(&state_count));
+    check_kern_return(kret, success);
+    regs.instruction_ptr = state.__rip as Address;
+    regs.stack_ptr = state.__rsp as Address;
+    regs.frame_ptr = state.__rbp as Address;
+}
+
+#[cfg(target_os = "windows")]
+#[allow(unsafe_code)]
+unsafe fn get_registers(thread_id: MonitoredThreadId, _regs: &mut Registers, _success: &mut bool) {
+    // TODO: use winapi::um::processthreadsapi::GetThreadContext
+    // and populate registers using the context.
+    // See https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-win32.cpp#129
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[allow(unsafe_code)]
+unsafe fn get_registers(_thread_id: MonitoredThreadId, _regs: &mut Registers, _success: &mut bool) {
     loop {
-        if BACKTRACE_READY.load(Ordering::SeqCst) {
+        if SAMPLEE_CONTEXT_READY.load(Ordering::SeqCst) {
             break;
         }
         thread::yield_now();
     }
-    let (thread_id, trace) = BACKTRACE.take().unwrap();
-    assert_eq!(thread_id, monitored.thread_id);
-    BACKTRACE_READY.store(false, Ordering::SeqCst);
-    Some(trace)
+    let (_thread_id, _context) = SAMPLEE_CONTEXT.take().unwrap();
+    // TODO: populate registers from context.
+    // See https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp#85
+}
+
+#[cfg(target_os = "macos")]
+#[allow(unsafe_code)]
+unsafe fn resume_thread(thread_id: MonitoredThreadId, success: &mut bool) {
+    check_kern_return(mach::thread_act::thread_resume(thread_id), success);
 }
 
 #[cfg(target_os = "windows")]
-unsafe fn get_backtrace_from_monitored_component(monitored: &MonitoredComponent) -> Option<Backtrace> {
-    None
+#[allow(unsafe_code)]
+unsafe fn resume_thread(thread_id: MonitoredThreadId, _success: &mut bool) {
+    // TODO: use winapi::um::processthreadsapi::ResumeThread
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+fn resume_thread(_thread_id: MonitoredThreadId, _success: &mut bool) {
+    SAMPLEE_MAY_RESUME.store(true, Ordering::SeqCst);
+}
+
+#[allow(unsafe_code)]
+unsafe fn frame_pointer_stack_walk(regs: &mut Registers, native_stack: &mut NativeStack) {
```

Comment:
Does regs need to be mutable?

Reply:
The reason I am passing all those `&mut` around as function argument, is because I wanted to avoid any potential allocation inside the "critical section". To be honest I have only limited understanding of the topic. 

So for example creating and returning a bool inside a function, would that not be doing a 'dynamic allocation'? And for those registers, I could just create them here and return the value, yet I also was "trying to prevent a dymamic allocation". 

I would off-course prefer not to pass around mutable references and instead just be returning values. Could you pls confirm that this would be ok to do in the "critical section", while the samplee thread is suspended?

Reply:
Dynamic allocation refers to anything that involves Vec, Box, Arc, Rc, etc. Anything that is stack allocated (such as a fixed-size array, or a bool, or a struct that doesn't contain types that require dynamic allocation) are totally fine to pass around and return.

Reply:
Great! thanks for the explanation!